File: directory_analyzer/rust-ver/src/output.rs
use std::path::Path;

use anyhow::Error;
use rust_xlsxwriter::{Format, Workbook, Worksheet};

use crate::{AnalyzedInfo, Args};

///Returns a string analyzed_info based on how args is configured
pub fn output(args: &Args, analyzed_info: Result<AnalyzedInfo, Error>) -> anyhow::Result<String> {
    match analyzed_info {
        Ok(info) => match args.export_xlsx() {
            Some(path) => {
                let workbook = create_workbook(&info)?;
                let path = write_workbook(path, workbook)?;
                let out = format!("Info saved to {}", &path.to_string_lossy());
                Ok(out)
            }
            None => Ok(format!("{info}")),
        },
        Err(e) => Err(e),
    }
}

fn write_workbook(path: &Path, mut workbook: Workbook) -> Result<std::path::PathBuf, Error> {
    let mut path = path.to_path_buf();
    if path.extension().is_none() {
        path.set_extension("xlsx");
    }
    workbook.save(&path)?;
    Ok(path)
}

fn create_workbook(info: &AnalyzedInfo) -> Result<Workbook, Error> {
    let mut workbook = Workbook::new();
    let bold_format = Format::new().set_bold();
    let mut worksheet = Worksheet::new();
    worksheet.set_name("General Info")?;
    let mut cur_col = 0;
    worksheet.write_with_format(0, cur_col, "Found dirs", &bold_format)?;
    worksheet.write(1, cur_col, info.found_dirs())?;
    cur_col += 1;
    worksheet.write_with_format(0, cur_col, "Found files", &bold_format)?;
    worksheet.write(1, cur_col, info.found_files())?;
    cur_col += 1;
    worksheet.write_with_format(0, cur_col, "Total bytes", &bold_format)?;
    worksheet.write(1, cur_col, info.total_bytes())?;
    cur_col += 1;
    if let Some(symlink_info) = info.found_symlinks() {
        worksheet.write_with_format(0, cur_col, "Found symlinks", &bold_format)?;
        worksheet.write(1, cur_col, symlink_info.found_symlinks())?;
        cur_col += 1;
        worksheet.write_with_format(0, cur_col, "File symlinks", &bold_format)?;
        worksheet.write(1, cur_col, symlink_info.file_symlinks())?;
        cur_col += 1;
        worksheet.write_with_format(0, cur_col, "Dir symlinks", &bold_format)?;
        worksheet.write(1, cur_col, symlink_info.dir_symlinks())?;
        // cur_col += 1;
    }
    worksheet.autofit();
    workbook.push_worksheet(worksheet);
    if let Some(file_info) = info.file_info() {
        let mut fi_worksheet = Worksheet::new();
        let mut cur_col = 0;
        fi_worksheet.set_name("File Info")?;
        let down_offset = 0;
        let tmp: Vec<_> = file_info.iter().enumerate().collect();
        fi_worksheet.write_with_format(down_offset, cur_col, "File type", &bold_format)?;
        cur_col += 1;
        fi_worksheet.write_with_format(down_offset, cur_col, "Num files", &bold_format)?;
        cur_col += 1;
        fi_worksheet.write_with_format(down_offset, cur_col, "% of total files", &bold_format)?;
        cur_col += 1;
        fi_worksheet.write_with_format(
            down_offset,
            cur_col,
            "Total size of files(bytes)",
            &bold_format,
        )?;
        cur_col += 1;

        fi_worksheet.write_with_format(down_offset, cur_col, "% of total bytes", &bold_format)?;
        cur_col += 1;

        fi_worksheet.write_with_format(down_offset, cur_col, "Largest file", &bold_format)?;
        cur_col += 1;
        fi_worksheet.write_with_format(
            down_offset,
            cur_col,
            "Largest file size(bytes)",
            &bold_format,
        )?;
        cur_col += 1;
        fi_worksheet.write_with_format(down_offset, cur_col, "Smallest file", &bold_format)?;
        cur_col += 1;
        fi_worksheet.write_with_format(
            down_offset,
            cur_col,
            "Smallest file size(bytes)",
            &bold_format,
        )?;

        for (i, (f_type, f_info)) in tmp {
            let i = i as u32 + down_offset + 1;
            let mut cur_col = 0;
            //File type
            fi_worksheet.write(i, cur_col, f_type)?;
            cur_col += 1;
            //Num Files
            fi_worksheet.write(i, cur_col, f_info.num_files())?;
            cur_col += 1;
            //% of total files
            let (per_tot_file, per_tot_size) = f_info.percentages_in_string();
            fi_worksheet.write(i, cur_col, per_tot_file)?;
            cur_col += 1;
            //Total size of files
            fi_worksheet.write(i, cur_col, f_info.size_in_bytes())?;
            cur_col += 1;
            //% of total size
            fi_worksheet.write(i, cur_col, per_tot_size)?;
            cur_col += 1;
            //Largest file path
            fi_worksheet.write(i, cur_col, f_info.largest_file().path().to_string_lossy())?;
            cur_col += 1;
            //Largest file size
            fi_worksheet.write(i, cur_col, f_info.largest_file().size())?;
            cur_col += 1;
            //Smallest file path
            fi_worksheet.write(i, cur_col, f_info.smallest_file().path().to_string_lossy())?;
            cur_col += 1;
            //Smallest file size
            fi_worksheet.write(i, cur_col, f_info.smallest_file().size())?;
        }
        fi_worksheet.autofit();
        workbook.push_worksheet(fi_worksheet);
    }
    Ok(workbook)
}

File: directory_analyzer/csharp-ver/dira/output.cs
using dira.types;
using ClosedXML.Excel;

namespace dira
{

    public static class B
    {
        public static string Output(Args args, AnalyzedInfo analyzedInfo)
        {
            if (args.ExportXLSX != null)
            {

                var workbook = CreateWorkbook(analyzedInfo);
                var path = WriteWorkbook(args.ExportXLSX, workbook);
                string format = "Info saved to {0}";
                return string.Format(format, path);

            }
            else
            {
                return string.Format("{0}", analyzedInfo);
            }
        }

        static XLWorkbook CreateWorkbook(AnalyzedInfo info)
        {
            var workbook = new XLWorkbook();
            var gen_worksheet = workbook.AddWorksheet("General Info");
            var cur_col = 1;
            gen_worksheet.Cell(1, cur_col).Value = "Found dirs";
            gen_worksheet.Cell(1, cur_col).Style.Font.SetBold();
            gen_worksheet.Cell(2, cur_col).Value = info.FoundDirs;
            gen_worksheet.Column(cur_col).AdjustToContents();
            cur_col += 1;
            gen_worksheet.Cell(1, cur_col).Value = "Found file";
            gen_worksheet.Cell(1, cur_col).Style.Font.SetBold();
            gen_worksheet.Cell(2, cur_col).Value = info.FoundFiles;
            gen_worksheet.Column(cur_col).AdjustToContents();
            cur_col += 1;
            gen_worksheet.Cell(1, cur_col).Value = "Total bytes";
            gen_worksheet.Cell(1, cur_col).Style.Font.SetBold();
            gen_worksheet.Cell(2, cur_col).Value = info.TotalBytes;
            gen_worksheet.Column(cur_col).AdjustToContents();
            cur_col += 1;
            if (info.FoundSymlinks is SymlinkInfo symlinkInfo)
            {
                gen_worksheet.Cell(1, cur_col).Value = "Found symlinks";
                gen_worksheet.Cell(1, cur_col).Style.Font.SetBold();
                gen_worksheet.Cell(2, cur_col).Value = symlinkInfo.FoundSymlinks;
                gen_worksheet.Column(cur_col).AdjustToContents();
                cur_col += 1;
                gen_worksheet.Cell(1, cur_col).Value = "File symlinks";
                gen_worksheet.Cell(1, cur_col).Style.Font.SetBold();
                gen_worksheet.Cell(2, cur_col).Value = symlinkInfo.GetFileSymlinks();
                gen_worksheet.Column(cur_col).AdjustToContents();
                cur_col += 1;
                gen_worksheet.Cell(1, cur_col).Value = "Dir symlinks";
                gen_worksheet.Cell(1, cur_col).Style.Font.SetBold();
                gen_worksheet.Cell(2, cur_col).Value = symlinkInfo.GetDirSymlinks();
                gen_worksheet.Column(cur_col).AdjustToContents();
            }

            if (info.FileInfo is Dictionary<FileExtension, FileTypeInfo> file_info)
            {
                var fi_worksheet = workbook.AddWorksheet("File Info");
                cur_col = 1;
                var down_offset = 1;
                fi_worksheet.Cell(down_offset, cur_col).Value = "File Type";
                fi_worksheet.Cell(down_offset, cur_col).Style.Font.SetBold();
                cur_col += 1;
                fi_worksheet.Cell(down_offset, cur_col).Value = "Num files";
                fi_worksheet.Cell(down_offset, cur_col).Style.Font.SetBold();
                cur_col += 1;
                fi_worksheet.Cell(down_offset, cur_col).Value = "% of total files";
                fi_worksheet.Cell(down_offset, cur_col).Style.Font.SetBold();
                cur_col += 1;
                fi_worksheet.Cell(down_offset, cur_col).Value = "Total size of files(bytes)";
                fi_worksheet.Cell(down_offset, cur_col).Style.Font.SetBold();
                cur_col += 1;
                fi_worksheet.Cell(down_offset, cur_col).Value = "% of total bytes";
                fi_worksheet.Cell(down_offset, cur_col).Style.Font.SetBold();
                cur_col += 1;
                fi_worksheet.Cell(down_offset, cur_col).Value = "Largest file";
                fi_worksheet.Cell(down_offset, cur_col).Style.Font.SetBold();
                cur_col += 1;
                fi_worksheet.Cell(down_offset, cur_col).Value = "Largest file size(bytes)";
                fi_worksheet.Cell(down_offset, cur_col).Style.Font.SetBold();
                cur_col += 1;
                fi_worksheet.Cell(down_offset, cur_col).Value = "Smallest file";
                fi_worksheet.Cell(down_offset, cur_col).Style.Font.SetBold();
                cur_col += 1;
                fi_worksheet.Cell(down_offset, cur_col).Value = "Smallest file size(bytes)";
                fi_worksheet.Cell(down_offset, cur_col).Style.Font.SetBold();
                var max_col = cur_col + 1;

                var tmp = file_info.ToArray();
                for (int i = 0; i < file_info.Count; i++)
                {
                    var k = i + 1 + down_offset;
                    cur_col = 1;
                    var f_info = tmp[i].Value;
                    //File type
                    fi_worksheet.Cell(k, cur_col).Value = tmp[i].Key.Ext;
                    cur_col += 1;
                    //Num Files
                    fi_worksheet.Cell(k, cur_col).Value = f_info.NumFiles;
                    cur_col += 1;
                    (var per_tot_file, var per_tot_size) = f_info.PercentagesInString();
                    //% of total files
                    fi_worksheet.Cell(k, cur_col).Value = per_tot_file.Percent;
                    cur_col += 1;
                    //Total size of files
                    fi_worksheet.Cell(k, cur_col).Value = f_info.SizeInBytes;
                    cur_col += 1;
                    //% of total size
                    fi_worksheet.Cell(k, cur_col).Value = per_tot_size.Percent;
                    cur_col += 1;
                    //Largest file path
                    fi_worksheet.Cell(k, cur_col).Value = f_info.LargestFile.Path;
                    cur_col += 1;
                    //Largest file size
                    fi_worksheet.Cell(k, cur_col).Value = f_info.LargestFile.Size;
                    cur_col += 1;
                    //Smallest file path
                    fi_worksheet.Cell(k, cur_col).Value = f_info.SmallestFile.Path;
                    cur_col += 1;
                    //Smallest file size
                    fi_worksheet.Cell(k, cur_col).Value = f_info.SmallestFile.Size;
                }

                foreach (var i in Enumerable.Range(1, max_col + 1))
                {
                    fi_worksheet.Column(i).AdjustToContents();
                }
            }

            return workbook;
        }

        static string WriteWorkbook(string path, XLWorkbook workbook)
        {
            var final_path = path;
            if (Path.GetExtension(path) != ".xlsx")
            {
                final_path += ".xlsx";
            }
            workbook.SaveAs(final_path);
            return final_path;
        }
    }
}
